// @theia/monaco-editor-core/esm/vs/editor/standalone/browser/standaloneLayoutService.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditorScopedLayoutService = void 0;
const dom = require("../../../base/browser/dom");
const event_1 = require("../../../base/common/event");
const layoutService_1 = require("../../../platform/layout/browser/layoutService");
const codeEditorService_1 = require("../../browser/services/codeEditorService");
const extensions_1 = require("../../../platform/instantiation/common/extensions");
let StandaloneLayoutService = class StandaloneLayoutService {
    constructor(_codeEditorService) {
        this._codeEditorService = _codeEditorService;
        this.onDidLayout = event_1.Event.None;
    }
    get dimension() {
        if (!this._dimension) {
            this._dimension = dom.getClientArea(window.document.body);
        }
        return this._dimension;
    }
    get hasContainer() {
        return false;
    }
    get container() {
        // On a page, multiple editors can be created. Therefore, there are multiple containers, not
        // just a single one. Please use `ICodeEditorService` to get the current focused code editor
        // and use its container if necessary. You can also instantiate `EditorScopedLayoutService`
        // which implements `ILayoutService` but is not a part of the service collection because
        // it is code editor instance specific.
        throw new Error(`ILayoutService.container is not available in the standalone editor!`);
    }
    focus() {
        var _a;
        (_a = this._codeEditorService.getFocusedCodeEditor()) === null || _a === void 0 ? void 0 : _a.focus();
    }
};
StandaloneLayoutService = __decorate([
    __param(0, codeEditorService_1.ICodeEditorService)
], StandaloneLayoutService);
let EditorScopedLayoutService = class EditorScopedLayoutService extends StandaloneLayoutService {
    constructor(_container, codeEditorService) {
        super(codeEditorService);
        this._container = _container;
    }
    get hasContainer() {
        return false;
    }
    get container() {
        return this._container;
    }
};
EditorScopedLayoutService = __decorate([
    __param(1, codeEditorService_1.ICodeEditorService)
], EditorScopedLayoutService);
exports.EditorScopedLayoutService = EditorScopedLayoutService;
(0, extensions_1.registerSingleton)(layoutService_1.ILayoutService, StandaloneLayoutService);
