// @theia/monaco-editor-core/esm/vs/editor/common/textModelGuides.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndentGuideHorizontalLine = exports.IndentGuide = exports.HorizontalGuidesState = void 0;
var HorizontalGuidesState;
(function (HorizontalGuidesState) {
    HorizontalGuidesState[HorizontalGuidesState["Disabled"] = 0] = "Disabled";
    HorizontalGuidesState[HorizontalGuidesState["EnabledForActive"] = 1] = "EnabledForActive";
    HorizontalGuidesState[HorizontalGuidesState["Enabled"] = 2] = "Enabled";
})(HorizontalGuidesState = exports.HorizontalGuidesState || (exports.HorizontalGuidesState = {}));
class IndentGuide {
    constructor(visibleColumn, className, 
    /**
     * If set, this indent guide is a horizontal guide (no vertical part).
     * It starts at visibleColumn and continues until endColumn.
    */
    horizontalLine) {
        this.visibleColumn = visibleColumn;
        this.className = className;
        this.horizontalLine = horizontalLine;
    }
}
exports.IndentGuide = IndentGuide;
class IndentGuideHorizontalLine {
    constructor(top, endColumn) {
        this.top = top;
        this.endColumn = endColumn;
    }
}
exports.IndentGuideHorizontalLine = IndentGuideHorizontalLine;
